src/ui/actions/timeline.ts

```diff
@@ -504,50 +492,77 @@ export function setPlaybackPrecachedTime(time: number): SetPlaybackPrecachedTime
   return { type: "set_playback_precached_time", time };
 }
 
-export function setFocusRegion(focusRegion: {
-  startTime: number;
-  endTime: number;
-}): SetFocusRegionAction {
-  return { type: "set_trim_region", focusRegion };
-}
-
-export function setFocusAroundTime(time: number, maxTime: number): UIThunkAction {
+export function setFocusRegion(focusRegion: FocusRegion | null): UIThunkAction {
   return (dispatch, getState) => {
     const state = getState();
-    const zoomRegion = getZoomRegion(state);
+    const currentTime = getCurrentTime(state);
 
-    const duration = zoomRegion.endTime;
-    const startTime = Math.max(time - maxTime / 2, 0);
-    const endTime = Math.min(time + maxTime / 2, duration);
+    // Stop playback (if we're playing) to avoid the currentTime from getting out of bounds.
+    const playback = getPlayback(state);
+    if (playback !== null) {
+      dispatch(stopPlayback());
+    }
 
-    dispatch(setFocusRegion({ endTime, startTime }));
-  };
-}
+    if (focusRegion !== null) {
+      const zoomRegion = getZoomRegion(state);
+      const { endTime: prevEndTime, startTime: prevStartTime } = getFocusRegion(state) || {};
 
-export function updateFocusRegion(operation: FocusOperation): UIThunkAction {
-  return (dispatch, getState) => {
-    const state = getState();
-    const hoverTime = getHoverTime(state)!;
-    const focusRegion = getFocusRegion(state)!;
-    const zoomRegion = getZoomRegion(state);
+      let { endTime, startTime } = focusRegion;
 
-    const duration = zoomRegion.endTime;
-    // To get around a handful of selection UI bugs, this limits the allowable trimmed
-    // region to 10% of the full duration of the replay.
-    const minRegion = duration * 0.1;
-    const { startTime, endTime } = focusRegion;
+      // Basic bounds check.
+      if (startTime < zoomRegion.startTime) {
+        startTime = zoomRegion.startTime;
+      }
+      if (endTime > zoomRegion.endTime) {
+        endTime = zoomRegion.endTime;
+      }
 
-    let value, type;
+      // Make sure our region is valid.
+      if (endTime < startTime) {
+        // If we need to adjust a dimension, it's the most intuitive to adjust the one that's being updated.
+        if (prevEndTime === endTime) {
+          startTime = endTime;
+        } else {
+          endTime = startTime;
+        }
+      }
+
+      // Make sure the current time stays within the bounds of our selected region.
+      if (currentTime < startTime) {
```

Comment:
Sure, I guess, but is that the Redux way of doing things?

Right now we potentially dispatch 3 actions:`setTimelineState`, `setTimelineToTime`, and then the focus region itself `"set_trim_region"`

1. The `setTimelineState` action is kind of a cheat because it just spreads in whatever props you pass it.
2. The `setTimelineToTime` action does other things (like `Video.seek`) so we can't skip it.
3. And then `"set_trim_region"` is what updates the focus region (the thing owned by this action).

We _could_ combine 1 and 3 into the same `setTimelineState` call but that isn't an obvious improvement to me. Let's tackle this with a follow up.